src/x86/runtime.rs

```diff
@@ -0,0 +1,120 @@
+    sse2 = 1,
+    /// SSE3 (Streaming SIMD Extensions 3)
+    sse3 = 2,
+    /// SSSE3 (Supplemental Streaming SIMD Extensions 3)
+    ssse3 = 3,
+    /// SSE4.1 (Streaming SIMD Extensions 4.1)
+    sse4_1 = 4,
+    /// SSE4.2 (Streaming SIMD Extensions 4.2)
+    sse4_2 = 5,
+    /// AVX (Advanced Vector Extensions)
+    avx = 6,
+    /// AVX2 (Advanced Vector Extensions 2)
+    avx2 = 7,
+    /// FMA (Fused Multiply Add)
+    fma = 8,
+    /// BMI1 (Bit Manipulation Instructions 1)
+    bmi1 = 9,
+    /// BMI1 (Bit Manipulation Instructions 2)
+    bmi2 = 10,
+    /// ABM (Advanced Bit Manipulation) on AMD / LZCNT (Leading Zero Count) on Intel
+    abm = 11,
+    /// TBM (Trailing Bit Manipulation)
+    tbm = 12,
+    /// POPCNT (Population Count)
+    popcnt = 13,
+
+    #[doc(hidden)]
+    __NonExhaustive
+}
+
+fn set_bit(x: u32, bit: u32) -> u32 {
+    debug_assert!(32 > bit);
+    x | 1 << bit
+}
+
+fn test_bit(x: u32, bit: u32) -> bool {
+    debug_assert!(32 > bit);
+    x & (1 << bit) != 0
+}
+
+fn inv_test_bit(v: u32, idx: u32) -> bool {
+    debug_assert!(32 > idx);
+    ((v >> idx) & 1) != 0
+}
+
+struct FeatureStorage(pub u32);
+
+impl FeatureStorage {
+    fn new() -> FeatureStorage {
+        let ebx;
+        let ecx;
+        let edx;
+
+        unsafe {
+            asm!("cpuid" : "={ecx}"(ecx), "={edx}"(edx)
+                 : "{eax}"(0x00000001u32), "{ecx}"(0 as u32) : :);
+            asm!("cpuid" : "={ebx}"(ebx) : "{eax}"(0x00000007u32), "{ecx}"(0 as u32) : :);
+        }
+
+        let mut value: u32 = 0;
+
+        // CPUID call with EAX=7, ECX=0 => Extended Features in EBX and ECX (unneeded):
+        if inv_test_bit(ebx, 3) { value = set_bit(value, __Feature::bmi1 as u32); }
+        if inv_test_bit(ebx, 5) { value = set_bit(value, __Feature::avx2 as u32); }
+        if inv_test_bit(ebx, 8) { value = set_bit(value, __Feature::bmi2 as u32); }
+
+        // CPUID call with EAX=1 => feature bits in ECX and EDX:
+        if inv_test_bit(ecx, 0) { value = set_bit(value, __Feature::sse3 as u32); }
+        if inv_test_bit(ecx, 5) { value = set_bit(value, __Feature::abm as u32); }
+        if inv_test_bit(ecx, 9) { value = set_bit(value, __Feature::ssse3 as u32); }
+        if inv_test_bit(ecx, 12) { value = set_bit(value, __Feature::fma as u32); }
+        if inv_test_bit(ecx, 19) { value = set_bit(value, __Feature::sse4_1 as u32); }
+        if inv_test_bit(ecx, 20) { value = set_bit(value, __Feature::sse4_2 as u32); }
+        if inv_test_bit(ecx, 21) { value = set_bit(value, __Feature::tbm as u32); }
+        if inv_test_bit(ecx, 23) { value = set_bit(value, __Feature::popcnt as u32); }
+        if inv_test_bit(ecx, 28) { value = set_bit(value, __Feature::avx as u32); }
+
+        if inv_test_bit(edx, 25) { value = set_bit(value, __Feature::sse as u32); }
+        if inv_test_bit(edx, 26) { value = set_bit(value, __Feature::sse2 as u32); }
+
+        FeatureStorage(value)
+    }
+}
+
+lazy_static! {
```

Comment:
Do we need `lazy_static!` here? Can we use [`std::sync::Once`](https://doc.rust-lang.org/std/sync/struct.Once.html) from the standard library so that we don't bring in any external deps?

Reply:
I'll look into it, never used `std::sync::Once` before. I just saw that lazy static solved the problem, is part of the nursery, and `no_std`.

Reply:
I think we'll actually want to eschew *everything* here and just use raw `AtomicUsize`, as that's the only thing that's actually no_std compatible. Hopefully shouldn't be too hard though! (just storing some bitflags globally)

Reply:
`AtomicUsize` sounds better. I've tried to search for the performance of an atomic load but haven't found much. Would it be worth it to try to make the load non-atomic?

Reply:
> just storing some bitflags globally

How do you distinguish "hasn't been filled yet" from "doesn't support it" (both of which seem like a 0)? Is the plan to have two atomics? Use a bit to denote "has been set"?

Reply:
> Would it be worth it to try to make the load non-atomic?

Nah, I wouldn't bother (plus I think this'd be unsound).

> Use a bit to denote "has been queried"?

Indeed!

Reply:
> Use a bit to denote "has been queried"?

I'll use `std::usize::MAX` as sentinel value.

Reply:
> Nah, I wouldn't bother

Ok, I was thinking about initializing this before `main`.

Reply:
I think we can just make it a plain `static mut FEATURES: usize` global. The worst thing that can happen is that some threads write to the variable without synchronization, but they will be writting the exact same value.

Reply:
Er no, that's undefined behavior. Let's stick to atomics, and only look to alternate strategies if that proves too slow.